win32ss/gdi/ntgdi/dibobj.c

```diff
@@ -1242,147 +1250,243 @@ NtGdiStretchDIBitsInternal(
-                                              pbmi,
-                                              dwUsage,
-                                              cjMaxBits,
-                                              cjMaxInfo,
-                                              TRUE,
-                                              hcmXform);
+ /* We must have a LPBITMAPINFO */
+    if (!pbmi)
+    {
+        DPRINT1("Error, Invalid Parameter.\n");
+        EngSetLastError(ERROR_INVALID_PARAMETER);
+        return 0;
     }
 
+    /* Test if we can READ the bitmap data */ 
+    _SEH2_TRY
+    {
+        ProbeForRead(pbmi, cjMaxInfo, 1);
+        if (DIB_GetBitmapInfo(&pbmi->bmiHeader, &width, &height, &planes, 
+            &bpp, &compr, &size) == -1)
+        {
+            DPRINT1("Invalid bitmap\n");
+            _SEH2_YIELD(goto cleanup1;)
+        }
+    }
+    _SEH2_EXCEPT(EXCEPTION_EXECUTE_HANDLER)
+    {
+        DPRINT1("Error, failed to read the DIB bits\n");
+        EngSetLastError(ERROR_INVALID_PARAMETER);
+        _SEH2_YIELD(goto cleanup1;)
+    }
+    _SEH2_END
+
     if (pjInit && (cjMaxBits > 0))
     {
-        pvBits = ExAllocatePoolWithTag(PagedPool, cjMaxBits, 'pmeT');
-        if (!pvBits)
+        safeBits = ExAllocatePoolWithTag(PagedPool, cjMaxBits, TAG_DIB);
+        if (!safeBits)
         {
             return 0;
         }
 
         _SEH2_TRY
         {
             ProbeForRead(pjInit, cjMaxBits, 1);
-            RtlCopyMemory(pvBits, pjInit, cjMaxBits);
+            RtlCopyMemory(safeBits, pjInit, cjMaxBits);
         }
         _SEH2_EXCEPT(EXCEPTION_EXECUTE_HANDLER)
         {
-            ExFreePoolWithTag(pvBits, 'pmeT');
+            ExFreePoolWithTag(safeBits, TAG_DIB);
             _SEH2_YIELD(return 0);
         }
         _SEH2_END
     }
     else
     {
-        pvBits = NULL;
+        safeBits = NULL;
     }
 
-    /* FIXME: Locking twice is cheesy, coord tranlation in UM will fix it */
-    if (!(pdc = DC_LockDc(hdc)))
+    /* Here we select between the dwRop with SRCCOPY or not. */
+    if (dwRop == SRCCOPY)
     {
-        DPRINT1("Could not lock dc\n");
-        EngSetLastError(ERROR_INVALID_HANDLE);
-        goto cleanup;
-    }
+        hdcMem = NtGdiCreateCompatibleDC(hdc);
+        if (hdcMem == NULL)
+        {
+            DPRINT1("NtGdiCreateCompatibleDC failed to create hdc.\n");
+            EngSetLastError(ERROR_NO_SYSTEM_RESOURCES);
+            return 0;
+        }
 
-    /* Calculate source and destination rect */
-    rcSrc.left = xSrc;
-    rcSrc.top = ySrc;
-    rcSrc.right = xSrc + abs(cxSrc);
-    rcSrc.bottom = ySrc + abs(cySrc);
-    rcDst.left = xDst;
-    rcDst.top = yDst;
-    rcDst.right = rcDst.left + cxDst;
-    rcDst.bottom = rcDst.top + cyDst;
-    IntLPtoDP(pdc, (POINTL*)&rcDst, 2);
-    RECTL_vOffsetRect(&rcDst, pdc->ptlDCOrig.x, pdc->ptlDCOrig.y);
+        hBitmap = NtGdiCreateCompatibleBitmap(hdc,
+                                              abs(pbmi->bmiHeader.biWidth),
```

Comment:
You can't use `pbmi` here and elsewhere, except in the SEH block to copy it or capture it's parameters, like done with `DIB_GetBitmapInfo`. So instead of using the `pbmi->xyz`fields, you would need to use `width`, `height`, `bpp`, `compr`. The problem remains at the end, where it's passed to `IntSeDIBits`, which will access it without SEH, and the function can also not be wrapped in SEH. So eventually the only correct solution is to copy `pbmi` to a new allocation.

Reply:
Doesn't the _SEH2_TRY at Line #1262 prove that pbmi is safe for us?

Reply:
No, because its validity can change under the hood (since it's a user-mode pointer).

Reply:
I have copied your code from the NtGdiGetDIBitsInternal function which should take care of this now.